pykern/pkresource.py

```diff
@@ -8,34 +8,82 @@
+import importlib
 import inspect
 import os.path
 import pkg_resources
 import re
 
+# TODO(e-carlin): discuss with rn if ok to import pkio
 from pykern import pkinspect
+from pykern import pkio
 from pykern import pksetup
 
 
-def filename(relative_filename, caller_context=None):
+def filename(relative_filename, caller_context=None, additional_packages=None, relpath=False):
     """Return the filename to the resource
 
     Args:
         relative_filename (str): file name relative to package_data directory.
         caller_context (object): Any object from which to get the `root_package`
+        additional_packages (List[str]): Packages to search first in addition to caller_module/caller_context
+        relpath (bool): If True path is relative to package package_data dir
 
     Returns:
-        str: absolute path of the resource file
+        str: absolute or relative (to package_data) path of the resource file
     """
-    pkg = pkinspect.root_package(
-        caller_context if caller_context else pkinspect.caller_module())
     assert not os.path.isabs(relative_filename), \
         'must not be an absolute file name={}'.format(relative_filename)
-    fn = os.path.join(pksetup.PACKAGE_DATA, relative_filename)
-    res = pkg_resources.resource_filename(pkg, fn)
-    if not os.path.exists(res):
-        msg = 'resource does not exist for pkg=' + pkg
-        if pkg == '__main__':
-            msg += '; do not call module as a program'
-        raise IOError((errno.ENOENT, msg, res))
+    a = []
+    for f, p in _files(relative_filename, caller_context, additional_packages):
+        a.append(p)
+        if os.path.exists(f):
+            if relpath:
+                f = str(pkio.py_path(
+                    os.path.join(pkg_resources.resource_filename(p, ''), pksetup.PACKAGE_DATA),
+                ).bestrelpath(pkio.py_path(f)))
+            return f
+    msg = f'unable to locate resource={relative_filename} in packages={a}'
+    if '__main__' in a:
+        msg += '; do not call module as a program'
+    raise IOError(errno.ENOENT, msg)
+
+
+def glob(relative_path, caller_context=None, additional_packages=None):
+    """Find all paths that match the relative path
+
+    Args:
+        relative_path(str): Path relative to package_data directory.
+        caller_context (object): Any object from which to get the `root_package`.
+        additional_packages (List[str]): Packages to search first in addition to caller_module/caller_context
+    Returns:
+        [str]: absolute paths of the matched files
+    """
+    res = []
+    for f, _ in _files(relative_path, caller_context, additional_packages):
+        res.extend(builtin_glob.glob(f))
     return res
+
+
+def _files(path, caller_context, additional_packages):
+    # Search additional_packages first so they can possibly override caller context/module
+    # files of the same name
+    for p in list(map(
+        lambda m: pkinspect.root_package(importlib.import_module(m)),
```

Comment:
It will be simpler if this is `search_path`.

Reply:
ccca032